mono/metadata/oop.c

```diff
@@ -0,0 +1,420 @@
+
+typedef gboolean(*ReadMemoryCallback)(void* buffer, gsize* read, const void* address, gsize size, void* userdata);
+typedef gboolean(*ReadExceptionCallback)(const void* address, gsize size, void* userdata);
+
+typedef struct _OutOfProcessMono
+{
+    ReadMemoryCallback readMemory;
+    ReadExceptionCallback readException;
+    void* userData;
+} OutOfProcessMono;
+
+static OutOfProcessMono g_oop = { NULL, NULL };
+
+#define OFFSET_MEMBER(type, base, member) ((gpointer)((gchar*)(base) + offsetof(type, member)))
+
+void read_exception(const void* address, gsize size)
+{
+    g_assert(g_oop.readException);
+    g_oop.readException(address, size, g_oop.userData);
+}
+
+gsize read_memory(void* buffer, const void* address, gsize size)
+{
+    if (!buffer || !size)
+        return 0;
+
+    gsize read = 0;
+    if (!g_oop.readMemory || !g_oop.readMemory(buffer, &read, address, size, g_oop.userData)) {
+        read_exception(address, size);
+    }
+    
+    return read;
+}
+
+// Read a null-terminated string out-of-process
+gsize read_nt_string(char* buffer, gsize max_size, const void* address)
+{
+    if (!buffer || !max_size)
+        return 0;
+
+    if (!g_oop.readMemory) {
+        read_exception(address, 1);
+        return 0;
+    }
+
+    gsize read = 0;
+    if (!g_oop.readMemory(buffer, &read, address, max_size, g_oop.userData)) {
+        // Failed to read, but just because we may not have read max_size, we still
+        // might be OK if at least one character was read (i.e. the null-terminator)
+        if (read == 0)
+            read_exception(address, 1);
+    }
+
+    // Ensure there's a null-terminator
+    buffer[min(read, max_size-1)] = '\0';
+
+    return read;
+}
+
+gpointer read_pointer(const void* address)
+{
+    gpointer ptr = NULL;
+    read_memory(&ptr, address, sizeof(ptr));
+    return ptr;
+}
+
+gint64 read_qword(const void* address)
+{
+    gint64 v = 0;
+    read_memory(&v, address, sizeof(v));
+    return v;
+}
+
+gint32 read_dword(const void* address)
+{
+    gint32 v = 0;
+    read_memory(&v, address, sizeof(v));
+    return v;
+}
+
+GList* read_glist_next(GList* list) { return (GList*) read_pointer(OFFSET_MEMBER(GList, list, next)); }
+gpointer read_glist_data(GList* list) { return read_pointer(OFFSET_MEMBER(GList, list, data)); }
+
+MONO_API void
+mono_unity_oop_init(
+    ReadMemoryCallback rmcb, 
+    ReadExceptionCallback recb,
+    void* userdata) 
+{
+    g_oop.readMemory = rmcb;
+    g_oop.readException = recb;
+    g_oop.userData = userdata;
+}
+
+MONO_API void
+mono_unity_oop_shutdown(void)
+{
+    memset(&g_oop, 0, sizeof(g_oop));
+}
+
+#ifdef _M_X64
+gboolean TryAcquireSpinWait(PSRWLOCK lock, unsigned int spinWait)
+{
+    do
+    {
+        if (TryAcquireSRWLockExclusive(&g_dynamic_function_table_lock))
+            return TRUE;
+    } while (spinWait--);
+
+    return FALSE;
+}
+#endif
+
+MONO_API GList*
+mono_unity_lock_dynamic_function_access_tables64(unsigned int spinWait) 
```

Comment:
I am confused about the locking. Is this occurring in the crashing process? Or in the inspection process?